app/src/main/java/com/android/wildex/ui/report/SubmitReportScreenViewModel.kt

```diff
@@ -38,4 +48,96 @@ class SubmitReportScreenViewModel(
     private val reportRepository: ReportRepository = RepositoryProvider.reportRepository,
     private val storageRepository: StorageRepository = RepositoryProvider.storageRepository,
     private val currentUserId: Id = Firebase.auth.uid ?: ""
-) : ViewModel() {}
+) : ViewModel() {
+  private val _uiState = MutableStateFlow(SubmitReportUiState())
+  val uiState: StateFlow<SubmitReportUiState> = _uiState.asStateFlow()
+
+  fun updateDescription(description: String) {
+    _uiState.value = _uiState.value.copy(description = description)
+  }
+
+  fun updateImage(imageUri: Uri?) {
+    _uiState.value = _uiState.value.copy(imageUri = imageUri)
+  }
+
+  fun clearErrorMsg() {
+    _uiState.value = _uiState.value.copy(errorMsg = null)
+  }
+
+  @SuppressLint("MissingPermission")
+  fun fetchUserLocation(locationClient: FusedLocationProviderClient) {
+    viewModelScope.launch {
+      try {
+        locationClient.lastLocation
+            .addOnSuccessListener { loc ->
+              if (loc != null) {
+                _uiState.value =
+                    _uiState.value.copy(
+                        location = Location(latitude = loc.latitude, longitude = loc.longitude))
+              } else {
+                setError("Unable to fetch current location.")
+              }
+            }
+            .addOnFailureListener {
+              setError("Failed to get location: ${it.message}. Please enable GPS.")
+            }
+      } catch (e: Exception) {
+        setError("Error fetching location: ${e.message}")
+      }
+    }
+  }
+
+  fun submitReport(onSuccess: () -> Unit) {
+    val currentState = _uiState.value
+
+    when {
+      currentState.imageUri == null -> {
+        setError("Please provide an image for the report.")
+        return
+      }
+      currentState.description.isBlank() -> {
+        setError("Please provide a description for the report.")
+        return
+      }
+      currentState.location == null -> {
+        setError("Location permission is required to submit a report.")
+        return
+      }
+    }
+
+    viewModelScope.launch {
+      try {
+        _uiState.value = currentState.copy(isSubmitting = true)
+        val authorId = currentUserId
+        val reportId = reportRepository.getNewReportId()
+        val imageUrl = storageRepository.uploadReportImage(reportId, currentState.imageUri!!)!!
```

Comment:
Reported by Sonar Cloud: the non-null assertion here is not needed, since you check it at line 94


```suggestion
        val imageUrl = storageRepository.uploadReportImage(reportId, currentState.imageUri)!!
```